### PR TITLE
[SPARK-41765][SQL] Pull out v1 write metrics to WriteFiles

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Pull out the metrics which is in `InsertIntoHiveTable` and `InsertIntoHadoopFsRelationCommand` to `WriteFiles`.

### Why are the changes needed?

Move metrics to the right place.

### Does this PR introduce _any_ user-facing change?

yes, the SQL UI metrics changed from `V1WriteCommand` to `WriteFiles`

with `spark.sql.optimizer.plannedWrite.enabled` disable and enable:

// disable
image

// enable
image

### How was this patch tested?

fix and improve test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/DataWritingCommand.scala [hunk 1]
```diff
@@ -79,27 +77,6 @@ object DataWritingCommand {
...
-   * [[DataWritingCommand]], but after running [[InsertIntoHadoopFsRelationCommand]]
-   * or [[InsertIntoHiveTable]], we only update metrics in these two command through
-   * [[BasicWriteJobStatsTracker]], we also need to propogate metrics to the command
-   * that actually calls [[InsertIntoHadoopFsRelationCommand]] or [[InsertIntoHiveTable]].
-   *
-   * @param sparkContext Current SparkContext.
-   * @param command Command to execute writing data.
-   * @param metrics Metrics of real DataWritingCommand.
-   */
-  def propogateMetrics(
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/DataWritingCommand.scala:94]: unused code

**Comment by ulysses-you** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
The commit related metrics should still belong to the insert command

**Comment by ulysses-you** [comment]
Seems task commit time is not related to insert command, driver side only do job commit. But if we split them, the UI will seem confused. That's why I moved both of them to `WriteFiles`. Any idea ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 2]
```diff
@@ -176,6 +153,40 @@ object FileFormatWriter extends Logging {
     //    V1 write command will be empty).
     if (Utils.isTesting) outputOrderingMatched = orderingMatched
 
+    SQLExecution.checkSQLExecutionId(sparkSession)
+
+    val finalStatsTrackers = if (writeFilesOpt.isDefined) {
+      val writeFilesMetrics = writeFilesOpt.get.metrics
+      val finalMetrics = statsTrackers match {
+        case Seq(tracker: BasicWriteJobStatsTracker) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:161]: should we use `statsTrackers.find`? in case there are more than one trackers in `statsTrackers` (seems possible w.r.t. the code)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala [hunk 3]
```diff
@@ -44,13 +48,14 @@ class BasicWriteJobStatsTrackerMetricSuite extends SparkFunSuite with LocalSpark
...
       // so the `statusStore.execution(executionId)` API is not used.
       assert(statusStore.executionsCount() == 2)
-      val executionData = statusStore.executionsList()(1)
-      val accumulatorIdOpt =
-        executionData.metrics.find(_.name == "number of dynamic part").map(_.accumulatorId)
-      assert(accumulatorIdOpt.isDefined)
-      val numPartsOpt = executionData.metricValues.get(accumulatorIdOpt.get)
-      assert(numPartsOpt.isDefined && numPartsOpt.get == partitions)
-
+      eventually(timeout(10.seconds)) {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala:51]: is this change related to this PR?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala [hunk 4]
```diff
@@ -57,6 +59,9 @@ case class InsertIntoHiveDirCommand(
     overwrite: Boolean,
     outputColumnNames: Seq[String]) extends SaveAsHiveFile with V1WritesHiveUtils {
 
+  override lazy val metrics: Map[String, SQLMetric] =
+    BasicWriteJobStatsTracker.metrics
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala:63]: should this be `writeFilesMetrics`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala (hunk 2 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:161]: It does not possible since all v1 write pass one tracker and we only do that if WriteFiles exists. But I'm fine to change it to be compatible with developers.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala (hunk 3 above)
Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala:51]: I think it is related due to the fact CI is failed, but it passed in my local.

I this the reason is, the metrics event lightly changed. Before all write file metrics come from event `SparkListenerDriverAccumUpdates`.  Now the metrics can be changed by both task event and `SparkListenerDriverAccumUpdates`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala:63]: Can we add a comment to explain it?

**Review by cloud-fan — APPROVED** [review]

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala (hunk 4 above)
Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala:65]: sure

**Comment by cloud-fan** [comment]
thanks, merging to master/3.4 (for the to-be-released planned write feature)!
